# Noise Cancelling Repeat Finder &mdash; Mock Genomes

Simulated pacbio and nanopore reads from fake genomes embedded with some tandem
repeats.

# Mock Genome 1

(to be described later)

# Mock Genome 2

(to be described later)

# Mock Genome 3

The genome is a mix of random sequence and tandem repeats. The primary repeat
elements are have periods of 5, 10, 20, 40, 80, and 171 bp, shown below. GGAAT
is a highly studied 5-mer, and the 171-mer is alpha satellite. The other four
are randomly generated sequences.

```bash  
GGAAT
CACTGCTGGT
GTTAGCGGTCGTGCTGATGG
GGCTCCTATCTCGCCTGTTCCCGGGTTCCTCTTATTCTCA
GAGATTGGAGTCCAAGAAATTCAGTCACCTTTCAGCGGTTCCAGTCACGGCGCTAAGTGCCTATTGACCCGCTACTGTTT
TCTGTCTAGTTTTTATATGAAGATATTCCCTTTTCCACCATAATCCTCAAAGCGCTCCAAATATCCACTTGCAGATTCTACAAAAAGAGTGTTTCCAAACTGCTCTATCAAAAGAAATGTTCAACTCTGTGAGTTGAATACACACATCACAAAGAAGTTTCTGAGAATGCT
```  

A total of 2,000 repeats were generated, with lengths sampled from a normal
distribution with mu=1000, sigma=400, but with minimum length 500. About half
of the repeats were chosen from the six primary elements. The others were from
elements with edit distance 1 from those; these elements are intended as false
positive bait for processes searching for the primary elements.

Repeats were positioned randomly across a single-sequence "genome", with a
minimum separation of 500 bp between neighboring repeats. A repeat can begin
anywhere within its repeated motif, and can appear as either forward or reverse
complement.

Total length of the genome is 3.3Mbp, of which 2.1Mbp is repeats.

The genome is in genome3.fa. Positions of the embedded motifs, in the genome,
are recorded in genome3.truth.dat.

Fake reads were sampled from this genome, separately simulating pacbio and
nanopore sequencing error profiles. The error profiles were derived from
published alignments. For pacbio this was 1.7% mismatch rate, 8.9% single-base
insertion, 4.3% single-base deletion. For nanopore, this was 7.4% mismatch
rate, 2.0% single-base insertion, 10.2% single-base deletion.

Overall coverage (of the genome) is intended to be about 5X. Read lengths were
pulled from a distribution with min=1K, max=20K, and mu=7,333 (the p.d.f. drops
linearly from a high at 1K to zero at 20K).

The reads are in genome3.pacbio.fa and genome3.nanopore.fa. The true alignments
of the reads (to the genome) are recorded in genome3.pacbio.cigars and
genome3.nanopore.cigars.

True alignments of the embedded motif intervals on the reads are recorded in
genome3.pacbio.truth.ncrf and genome3.nanopore.truth.ncrf. Note that while
these are in a format similar to the output from NcRF, they were not generated
by an aligner. Instead, they are derived from the truth that was generated by
the simulation process.

# Mock Genome 4

The genome is similar in concept to mock genome 3. The difference is that
the repeat lengths are longer for longer motifs (in mock genome 3 the repeat
length distribution was the same, regardless of motif length). The intent is
that the distribution of copy numbers will be consistent regardless of motif
length.

The distribution of repeat lengths for the 5-mer is the same as in mock genome
3. The distributions for the other motifs are shown in this table.s
```bash  
5-mer    mu=1000   sigma=400   minLen=500
10-mer   mu=2000   sigma=800   minLen=1000
20-mer   mu=4000   sigma=1600  minLen=2000
40-mer   mu=8000   sigma=3200  minLen=4000
80-mer   mu=16000  sigma=6400  minLen=8000
171-mer  mu=34200  sigma=13680 minLen=17100
```  

The repeats are, on average, much longer; therefor the genome is much longer.
The coverage and read length distribution were kept the same, thus the number
of reads is much larger. Because of this increase in file size, all the files
are included in gzipped form.

# Truth File Formats

When a mock genome is generated, the positions of the embedded repeat arrays
*in the genome* are recorded in genome_name.truth.dat.  This is a
whitespace-delimited text file with seven columns. The first line is a header
line, beginning with a # and providing the names of the columns.

The columns are
```bash  
chrom,start,end  The interval containing the repeat array. Chrom is usually
                 the name of the genome. Start and end are origin-zero, half
                 open.
motif            The repeated motif. This has an optional suffix describing
                 the offset and orientation of the motif. For example,
                 "GGAAT.2+" indicates the array begins with the 3rd character
                 (the first A) and is on the forward strand.
rptLen           The number of bases in the repeat array.
len              The length of the interval containing the array (the same as
                 end-start). This will usually be identical to rptLen, but
                 allows for the possibility that some noise process has changed
                 the length of the array before it was embedded in the genome.
fill             The length of random sequence inserted between this repeat
                 array and the closest one that precedes it.
```  

When simulated reads are pulled from a mock genome is generated, the positions
of the embedded repeat arrays *in the reads* are recorded in
genome_name.technology_name.truth.dat. This is a whitespace-delimited text
file with seven columns. There is no header line.

The columns are
```bash  
chrom,gStart,gEnd  The interval *on the genome* containing the portion of the
                   repeat array that was sampled by this read. Start and end
                   are origin-zero, half open.
read,rStart,rEnd   The interval *on the read* corresponding to the interval on
                   the genome. This will usually have a different length than
                   the interval on the genome, because of simulated sequencing
                   noise.
motif              The repeated motif. This has an optional suffix describing
                   the orientation of the motif *on the read*.
```  
