# Noise Cancelling Repeat Finder &mdash; Mock Genomes

Simulated PacBio and Oxford Nanopore reads from fake genomes embedded with some
tandem repeats.

_For historical reasons, and to avoid having to change many development
scripts, the only mock genome is called "mock genome 3". No other mock genomes
are currently extant._


# Mock Genome 3

The genome is a mix of random sequence and tandem repeats. The primary repeat
elements have periods of 5, 10, 20, 40, 80, and 171 bp, shown below. GGAAT is a
highly studied 5-mer, and the 171-mer is alpha satellite. The other four are
randomly generated sequences.

These are the six primary repeat elements:

```bash  
GGAAT
CACTGCTGGT
GTTAGCGGTCGTGCTGATGG
GGCTCCTATCTCGCCTGTTCCCGGGTTCCTCTTATTCTCA
GAGATTGGAGTCCAAGAAATTCAGTCACCTTTCAGCGGTTCCAGTCACGGCGCTAAGTGCCTATTGACCCGCTACTGTTT
TCTGTCTAGTTTTTATATGAAGATATTCCCTTTTCCACCATAATCCTCAAAGCGCTCCAAATATCCACTTGCAGATTCTACAAAAAGAGTGTTTCCAAACTGCTCTATCAAAAGAAATGTTCAACTCTGTGAGTTGAATACACACATCACAAAGAAGTTTCTGAGAATGCT
```  

A total of 2,000 repeats were generated, with lengths sampled from a normal
distribution with mu=1000, sigma=400, but with minimum length 500. About half
of the repeats were chosen from the six primary elements. The others were from
elements with edit distance 1 from those -- these elements are intended as
false positive bait for processes searching for the primary elements.

Repeats were positioned randomly across a single-sequence "genome", with a
minimum separation of 500 bp between neighboring repeats. A repeat can begin
anywhere within its repeated motif, and can appear as either forward or reverse
complement.

Total length of the genome is 3.3Mbp, of which 2.1Mbp is repeats.

The genome is in genome3.fa. Positions of the embedded motifs, in the genome,
are recorded in genome3.truth.dat.

Fake reads were sampled from this genome, separately simulating PacBio and
Oxford Nanopore sequencing error profiles. The error profiles were derived from
published alignments (details are, or will be, in the manuscript). For PacBio
this was 1.7% mismatch rate, 8.9% single-base insertion, 4.3% single-base
deletion. For Oxford Nanopore, this was 4.6% mismatch rate, 3.8% single-base
insertion, 7.7% single-base deletion.

Overall coverage (of the genome) is intended to be about 5X. Read lengths were
pulled from a distribution with min=1K, max=20K, and mu=7,333 (the p.d.f. drops
linearly from a high at 1K to zero at 20K). This distribution was chosen as a
crude match for commonality of observed length distributions in published reads
for both PacBio and Oxford Nanopore -- though both technologies can produce
reads up to ten times this length, such reads are a very small part of the
observed distributions.

The simulated reads are in genome3.pacbio.fa and genome3.nanopore.fa. True
alignments of the reads, to the genome, are recorded in genome3.pacbio.cigars
and genome3.nanopore.cigars.

True alignments of the embedded motif intervals on the reads are recorded in
genome3.pacbio.truth.ncrf and genome3.nanopore.truth.ncrf, and more consisely
in genome3.pacbio.truth.summary and genome3.nanopore.truth.summary. *Note that
while these are in a format similar to the output from NcRF, they were not
generated by an aligner.* Instead, they are derived from the truth that was
generated by the simulation process.

# Truth File Formats

When a mock genome is generated, the positions of the embedded repeat arrays
*in the genome* are recorded in genome_name.truth.dat. This is a
whitespace-delimited text file with seven columns. The first line is a header
line, beginning with a waffle (#) and providing the names of the columns.

The columns are
```bash  
chrom,start,end    The interval containing the repeat array. Chrom is usually
                   the name of the genome. Start and end are origin-zero, half
                   open.
motif              The repeated motif. This has an optional suffix describing
                   the offset and orientation of the motif. For example,
                   "GGAAT.2+" indicates the array begins with the 3rd character
                   (the first A) and is on the forward strand.
rptLen             The number of bases in the repeat array.
len                The length of the interval containing the array (the same as
                   end-start). This will usually be identical to rptLen, but
                   allows for the possibility that some noise process has
                   changed the length of the array before it was embedded in
                   the genome.
fill               The length of random sequence inserted between this repeat
                   array and the closest one that precedes it.
```  

When simulated reads are pulled from a mock genome, the true alignments of the
reads to the genome are recorded in genome_name.technology_name.cigars. This
information is then used to recover the true alignments of the embedded repeat
arrays *in the reads* (discussed below). This is a whitespace-delimited text
file with five columns. There is no header line.

The columns are
```bash  
read               The name of the read.
chrom,gStart,gEnd  The interval ON THE GENOME from which the read was sampled.
                   This has an optional suffix describing the orientation of
                   the read relative to the genome. For example, "genome3-"
                   indicates the read was sampled from what would be the
                   genome's reverse strand. gStart and gEnd are origin-zero,
                   half open.
cigar              Cigar string describing the alignment. "M" indicates base(s)
                   in the read matching the genome, or mismatching (a
                   substitution). "I" indicates base(s) in the read that are
                   not in the genome. "D" indicates base(s) in the genome that
                   are not in the read.
```  

The true alignments of the embedded repeat arrays *in the reads* are recorded
in genome_name.technology_name.truth.ncrf. This is similar to the alignment
format produced by NCRF (described in the main README), with two fields
repurposed. First, where the motif and strand would normally be indicated,
.truth.ncrf files provide the aligned interval and orientation on the genome.
Second, the score field has no meaning in .truth.ncrf files.

Alignments are separated from each other by blank lines. A typical alignment
block is shown below. Note that there may be other lines in a block, each
starting with a waffle (#).
```bash  
# score=1 querybp=539 mRatio=87.6% m=507 mm=7 i=40 d=25 =======x=============...
genome3_PACBIO_F_2560_12833     10757 554bp   3756-4310 GGAATGGGAATGGAATGGAAT...
genome3:6143-6682+                    539bp   score=1   GGAATGG-AATGGAATGGAAT...
```  

The true positions of the embedded repeat arrays *in the reads* are recorded in
genome_name.technology_name.truth.summary. This is nearly identical to the
alignment summary format produced by ncrf_summary, but column 1 is completely
different and the meaning of some of the other columns is also different. This
is a whitespace-delimited text file with thirteen columns. The first line is a
header line, beginning with a waffle (#) and providing the names of the columns.

The columns are
```bash  
genome             The aligned interval ON THE GENOME containing the portion of
                   the repeat array. This is of the form chrom:gStart-gEnd.
                   gStart and gEnd are origin-zero, half open. For example,
                   "genome3:6143-6682" indicates an interval of 539 bp on
                   genome 3. 
motif              The embedded repeat, e.g. "GGAAT".
seq,start,end      The aligned interval ON THE READ (seq is the read name).
strand             "+" indicates the aligned intervals are in the same
                   orientation; "-" indicates they are in opposing orientations.
seqLen             The length of the read (the entire read, not just the
                   aligned interval).
querybp            The number of bases in the aligned repeat. This should be
                   the same as gEnd-gStart.
mRatio             The ratio of matches to alignment events, as a percentage,
                   e.g. "85.7%". This is computed as m/(m+mm+i+d).
m                  The number of matched bases in the alignment.
mm                 The number of mismatched bases in the alignment.
i                  The number of inserted bases in the alignment -- bases in
                   the read which are not in the genome.
d                  The number of deleted bases in the alignment -- bases in
                   the genome which are not in the read.
```  
